### PR TITLE
Add pyramid output for densenet, cspDarknet

### PR DESCRIPTION
**Pull request by sachinprasadhs**
- [x] Change from `Backbone` class to `FeaturePyramidBackbone` to get `pyramid outputs` for DenseNet and CSPdarkNet
- [x] Make common test case for pyramid outputs.

### CODE REVIEW

File: keras_nlp/src/models/csp_darknet/csp_darknet_backbone.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ class CSPDarkNetBackbone(Backbone):
             `"basic_block"` for basic conv block.
             Defaults to "basic_block".
         image_shape: tuple. The input shape without the batch size.
-            Defaults to `(None, None, 3)`.
```
Comment (mattdangerw) [inline keras_nlp/src/models/csp_darknet/csp_darknet_backbone.py:42]: actually I think this is what we want, as long as the backbone does actually support different sized inputs. so maybe update the default below?

File: keras_nlp/src/tests/test_case.py [hunk 2]
```diff
@@ -604,5 +604,27 @@ def compare(actual, expected):
 
             tree.map_structure(compare, output, expected_partial_output)
 
+    def run_pyramid_output_test(
```
Comment (mattdangerw) [inline keras_nlp/src/tests/test_case.py:607]: Can we just add this to the vision backbone test above? `expected_pyramid_output_keys=None` and `if expected_pyramid_output_keys` run everything in this block?

Would make our tests shorter.

File: keras_nlp/src/tests/test_case.py [hunk 3]
```diff
@@ -604,5 +604,27 @@ def compare(actual, expected):
...
+        output_data = model(input_data)
+
+        self.assertIsInstance(output_data, dict)
+        self.assertEqual(
+            list(output_data.keys()), list(backbone.pyramid_outputs.keys())
+        )
+        self.assertEqual(list(output_data.keys()), expected_pyramid_output_keys)
+
+        for k, v in output_data.items():
+            size = input_data.shape[1] // (2 ** int(k[1:]))
```
Comment (mattdangerw) [inline keras_nlp/src/tests/test_case.py:626]: Is this part true for all feature pyramid outputs? Or only some?

**Review by mattdangerw — COMMENTED** [review]
Just a couple comments.

File: keras_nlp/src/tests/test_case.py (hunk 3 above)
Comment (sachinprasadhs) [reply keras_nlp/src/tests/test_case.py:626]: This is true for Resnet, Densenet, CSPdarknet. 
For Mix transformer it will be with the slight change `size = input_data.shape[1] // (2 ** (int(k[1:])+1))` which does not fit into this testcase.

Comment (sachinprasadhs) [reply keras_nlp/src/tests/test_case.py:626]: And When I checked, this holds good for all efficientNet networks, mobilenet as well.

File: keras_nlp/src/tests/test_case.py (hunk 2 above)
Comment (sachinprasadhs) [reply keras_nlp/src/tests/test_case.py:622]: Done

File: keras_nlp/src/tests/test_case.py [hunk 4]
```diff
@@ -515,6 +523,13 @@ def run_vision_backbone_test(
                 run_mixed_precision_check=run_mixed_precision_check,
                 run_quantization_check=run_quantization_check,
             )
+            if expected_pyramid_output_keys:
```
Comment (mattdangerw) [inline keras_nlp/src/tests/test_case.py:526]: let's just run this once above, doubling up might inflate our test sizes too much (a perennial problem)

File: keras_nlp/src/tests/test_case.py [hunk 5]
```diff
@@ -491,6 +492,13 @@ def run_vision_backbone_test(
             run_mixed_precision_check=run_mixed_precision_check,
             run_quantization_check=run_quantization_check,
         )
+        if expected_pyramid_output_keys:
+            self.run_pyramid_output_test(
```
Comment (mattdangerw) [inline keras_nlp/src/tests/test_case.py:496]: I'd just move the function directly in to this test case.

File: keras_nlp/src/tests/test_case.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/src/tests/test_case.py:641]: Got it. Why don't we write this in a way that would be really general and obvious to the reader. Something like this in the actual test case. Then here we can just assert that `shape[1:3]` is the expected image size.

```
expected_pyramid_output_keys=["P2", "P3", "P4", "P5"],
expected_pyramid_image_sizes=[(128, 128), (64, 64), (32, 32), (16, 16)],
```

File: keras_nlp/src/tests/test_case.py [hunk 6]
```diff
@@ -604,5 +619,27 @@ def compare(actual, expected):
...
+
+        self.assertIsInstance(output_data, dict)
+        self.assertEqual(
+            list(output_data.keys()), list(backbone.pyramid_outputs.keys())
+        )
+        self.assertEqual(list(output_data.keys()), expected_pyramid_output_keys)
+
+        for k, v in output_data.items():
+            size = input_data.shape[1] // (2 ** int(k[1:]))
+            self.assertEqual(tuple(v.shape[:3]), (2, size, size))
```
Comment (mattdangerw) [inline keras_nlp/src/tests/test_case.py:642]: does this assume the batch size is 2? I think that is a bad general assumption, might just be true today. comment above would handle this though

**Review by mattdangerw — COMMENTED** [review]
Thanks! A few more comments on the test case work.

File: keras_nlp/src/tests/test_case.py (hunk 3 above)
Comment (sachinprasadhs) [reply keras_nlp/src/tests/test_case.py]: Changed, Thanks

File: keras_nlp/src/tests/test_case.py (hunk 6 above)
Comment (sachinprasadhs) [reply keras_nlp/src/tests/test_case.py]: Changed to check only `height` and `width` since we are checking the `batch_size` in the `run_backbone_basics` test.

**Review by mattdangerw — APPROVED** [review]
LGTM! Thanks!
